### ISSUE TITLE
OPENAI_BASE_URL instead of OPENAI_API_HOST

### ISSUE THREAD

**Issue by johnd0e** [mountaineerbr/shellChatGPT#3]
Right now `/v1` is part of ENDPOINT.
But it is [more common to have BASE_URL](https://github.com/openai/openai-python?tab=readme-ov-file#configuring-the-http-client) (or API_BASE) setting, which include `/v1` part as well: `https://api.openai.com/v1`.
Btw, this is so for GOOGLE_API_HOST, but not for others.

The reason: there are some services, which do not contain `/v1` part at all
E.g. `https://models.inference.ai.azure.com/chat/completions`

Ref: https://docs.github.com/en/github-models/prototyping-with-ai-models

**Comment by mountaineerbr** [comment]
That is great you brought up this subject. Indeed, I have struggled to understand how to try and make this feature the most compatible across multiple service providers. I will check if we can make this naming convention more correct and also check if we can have `/v1` moved to `API_HOST` variable... However, I remember to have tried doing this once but soon gave up because it would need changing lots of code instances... Thank you so much for clarifying this issue and giving me some ref pointer as I was completely lost about how to implement this right!

**Comment by mountaineerbr** [comment]
It seems to be possible to correct both these issues without too many changes in the code of the script at this time...

Even though we have updated the script to solve these issues today, we need some testing before release. I also found some more bugs (minor bugs) which are fixed now, however I need to test at least all the seven service provider configurations and functions before public release.

I am not sure if I can create an account to use with Azure yet, but I will see it in the next few days so I can test your example server. In any case, I will do my best to make this script more compatible with other services out-of-the-box. 
Cheers!

**Comment by johnd0e** [comment]
> am not sure if I can create an account to use with Azure yet

It is github account. But you need to create special token.

**Comment by mountaineerbr** [comment]
Doing some research, according to [ref1](https://swagger.io/docs/specification/v2_0/api-host-and-base-path/) and [ref2](https://www.ibm.com/docs/en/api-connect/saas?topic=definition-specifying-basepath-api), **baseUrl** in our case would be exactly `/v1`.

This means `API_HOST` points to the provider website root address, while `BASE_URL` would mean `/v1`.

As I tested `chatgpt.sh` script, it seems some service providers just ignore the `/v1` part of the url path, so it even works with Mistral and GitHub Models APIs. For now!

As our script targets [OpenAI API **version 1**](https://github.com/mountaineerbr/shellChatGPT#%EF%B8%8F-limitations), I reckon we shall not be fixing this seemingly issue.. If unlikely the project decides to move on and support a future `/v2` past version 1 of OpenAI API, we may reconsider this issue ticket!

As for records, the script in the present version (v0.81.0) already works with GitHub Models. The syntax for it to work with GitHub Models would be:

```
OPENAI_API_HOST="https://models.inference.ai.azure.com" chatgpt.sh -cc -m Phi-3-small-8k-instruct --api-key $GITHUB_TOKEN
```


The _workaround_ (which already exists) for the script to work with a service provider which fails when `/v1` is set automatically is to set the following envar instead:

```
OPENAI_API_HOST_STATIC="https://models.inference.ai.azure.com/chat/completions"
```

The `STATIC` parameter disables the script auto-selection of endpoints!

Lastly, if the user creates a **configuration file** from the template (`option -FF`), there are examples of how to set these parameters in it! And also, it is possible to wrap things even further with bash aliases which may define different configuration files for each use case (using envars for the config filepath in this case), if hard-setting these variables in the configuration file stands in the way of the user using more than one service provider!

---

The syntaxes above are described in the man page and in the Repository README.md.

Therefore, it is unlikely that this issue ticket is going to be addresses because I don't see it as a real issue in our code context _at the moment_.

Even though, I note that there is around ~4 times more code mentions of `OPENAI_BASE_URL` than `OPENAI_API_HOST` when I search these terms in GitHub searching mechanism...


---

**Most fortunately**, I did find that the GitHub Models service is great and so I went ahead and implemented the command line option `--github`, which should set up the script. With this, available models are listed with `option -l` and the current GitHub Models maximum token capacity are set correctly!

Just remember to have `$GITHUB_TOKEN` set and exported as an environment variable!

This is going to be available in **version v0.81.1** which shall be released when I finish testing the new API service provider...

**Comment by mountaineerbr** [comment]
Just in case you can help me, do you know any reliable way to dump GitHub Models as a list? I am using this for now:


```
curl "https://github.com/marketplace/models" |
sed -n 's/"original_name":"[^"]*",/\n&\n/gp' | sed -n 's/"original_name"://p' | sed 's/[",]//g'
```


**And ALSO**, when we try the `/info` endpoint and run:

```
curl https://models.inference.ai.azure.com/info\?model=Phi-3-small-8k-instruct -X POST -H 'Content-Type: application/json' -H 'Authorization: Bearer xoxoxoxo'
```

It returns:

```
{"error":{"code":"no_model_name","message":"No model specified in request. Please provide a model name in the request body or as a x-ms-model-mesh-model-name header.","details":null}}
```

**Comment by johnd0e** [comment]
```sh
curl -s https://models.inference.ai.azure.com/models \
    -H "Authorization: Bearer github_pat_xxx" \
   | jq .[].name
```

```
AI21-Jamba-Instruct
Cohere-command-r
Cohere-command-r-plus
Cohere-embed-v3-english
Cohere-embed-v3-multilingual
Meta-Llama-3-70B-Instruct
Meta-Llama-3-8B-Instruct
Meta-Llama-3.1-405B-Instruct
Meta-Llama-3.1-70B-Instruct
Meta-Llama-3.1-8B-Instruct
Mistral-large
Mistral-large-2407
Mistral-Nemo
Mistral-small
gpt-4o
gpt-4o-mini
text-embedding-3-large
text-embedding-3-small
Phi-3-medium-128k-instruct
Phi-3-medium-4k-instruct
Phi-3-mini-128k-instruct
Phi-3-mini-4k-instruct
Phi-3-small-128k-instruct
Phi-3-small-8k-instruct
Phi-3.5-mini-instruct
```

**Comment by johnd0e** [comment]
> This means `API_HOST` points to the provider website root address, while `BASE_URL` would mean `/v1`.

It's not about proper terminology.

What I'm trying to convey is that most projects use https://api.openai.com/v1 as the configurable part that defines the API. There are real APIs that do not include the v1 part; some may use v2. This is not an issue, as all these cases can be addressed with a single option.

Of course, there's no practical reason to customize the host and base separately. Therefore, the main reason for continuing to use the current API_HOST setting is compatibility, to avoid disrupting existing user setups.

> Therefore, it is unlikely that this issue ticket is going to be addresses because I don't see it as a real issue in our code context at the moment.

Well, you are the boss.

**Comment by mountaineerbr** [comment]
I really hope this project remains stable before `/v2`! It is too much code,and the AI models aren't that really worth the effort sometimes...

**Comment by johnd0e** [comment]
> I really hope this project remains stable before /v2

It's not about api.openai.com compatibility.
It is about support of 3rd-party "openai-compatible" APIs, which may have different base URLs right now.
I have seen even '/v1/openai' for example.

**Comment by mountaineerbr** [comment]
Hey, I just released the script with GitHub Models integration!

The API call to list models you suggested does not show all 35/36 models available, so I combined the two methods (name scraping + API call) to list all the available models in the terminal. See [the readme.md](https://github.com/mountaineerbr/shellChatGPT#github).

There remains more testing to be done, but I reckon it is usable for now (with vision support!).

Thank you so much for raising this issue. As I said before, we may reconsider this issue in the future if it proves to be necessary.

I also mentioned your account name in our [contributor list](https://github.com/mountaineerbr/shellChatGPT#-contributors)!
Cheers!

**Comment by johnd0e** [comment]
> As I said before, we may reconsider this issue in the future if it proves to be necessary.

For the record:

- https://api.novita.ai/v3/openai
- https://api.deepinfra.com/v1/openai

**Comment by johnd0e** [comment]
> The API call to list models you suggested does not show all 35/36 models available

So extra:

AI21-Jamba-1.5-Large
AI21-Jamba-1.5-Mini
Cohere-command-r-08-2024
Cohere-command-r-plus-08-2024
Ministral-3B

Do Cohere*08-2024 work for you?

**Comment by mountaineerbr** [comment]
> For the record:
> 
>     * https://api.novita.ai/v3/openai
> 
>     * https://api.deepinfra.com/v1/openai

OK, this makes your point even more valid and I can see what you mean now more clearly. I need to do some more research to feel comfortable with this proposal (I am a biologist!). Changing the script code to `URL_BASE` is not too hard, and I do wish this script to become the most compatible with other API services..

---

The `Cohere*08-2024` models do not work for me:

```
"message": "The model `Cohere-command-r-plus-08-2024` does not exist or you do not have access to it.",
```

Good catch! I did not have time to test all GitHub Models...

---

PS: The following two projects which we have as reference [TheR1D's shell_gpt](https://github.com/TheR1D/shell_gpt/) and [xenodium's chatgpt-shell](https://github.com/xenodium/chatgpt-shell) do use `BASE_URL`, so  I guess you are correct in stating that this is the way to go... 

However, upon close examination, xenodium's code:

```
(setq chatgpt-shell-api-url-base "https://{your-resource-name}.openai.azure.com")
...
((chatgpt-shell-api-url-base "https://api.chatgpt.domain.com")
```

TheRD1's:

```
ENV API_BASE_URL=http://10.10.10.10:11434
```

and [Mudler's LocalAI](https://github.com/mudler/LocalAI):

```
base_path = os.environ.get('OPENAI_API_BASE', 'http://localhost:8080/v1')
....
base_url="http://localhost:8080/v1/"
```


Which makes me confused, but I shall look further into this when I have got more time! Cheers!

**Comment by johnd0e** [comment]
For the reference
- OPENAI_API_BASE - the env variable that were used in OpenAI's python sdk (old persions)
- OPENAI_BASE_URL - the new name for the same variable, is used in new versions of their sdk

But in both cases default is `https://api.openai.com/v1`

**Comment by mountaineerbr** [comment]
This issue is addressed in [v0.83 -- GitHub commit](https://github.com/mountaineerbr/shellChatGPT/commit/bf5c49d64e79dbfb82a8e57de1557103052e4e68).

Thanks again for all the explanation and pointers!
